### ISSUE TITLE
Rust-Analyzer with Clippy enabled doesn't highlight errors in Cargo Workspaces with many empty crates

### ISSUE THREAD

**Issue by Ferdi265** [rust-analyzer/rust-analyzer#7096]
When working in a Cargo Workspace with many crates (>=6), most of which are still empty, I noticed that errors were not displayed at all in the crate I was working on when using Clippy as the `checkOnSave` command. The same does not happen when using Check instead of Clippy.

I expect that part of the reason for this is that Clippy errors out on the `it_works` testcase in the default new crate template, since it contains `assert_eq!(2 + 2, 4)`, which triggers a default-deny Clippy lint.

However, what was unexpected is that errors as simple as `let x = y;` (with no prior definition of `y`) are not recognized in the non-empty crates when too many crates are still "defaulted".

Example project tree:
```
.
├── Cargo.toml
├── Cargo.lock
├── bug (depends on dep1)
│   ├── Cargo.toml
│   └── src
│       └── lib.rs (trivial `let x = y;` error that is not detected)
├── dep1 (depends on dep2)
│   ├── Cargo.toml
│   └── src
│       └── lib.rs (default content)
├── dep2
│   ├── Cargo.toml
│   └── src
│       └── lib.rs (default content)
├── nop1
│   ├── Cargo.toml
│   └── src
│       └── lib.rs (default content)
├── nop2
│   ├── Cargo.toml
│   └── src
│       └── lib.rs (default content)
└── nop3
    ├── Cargo.toml
    └── src
        └── lib.rs (default content)
```

Contents of `bug/src/lib.rs`:
```rust
fn wut() {
    let x = y;
}
```

Contents of all other `lib.rs` files:
```rust
#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        assert_eq!(2 + 2, 4);
    }
}
```

Test Workspace: [workspace.tar.gz](https://github.com/rust-analyzer/rust-analyzer/files/5755010/workspace.tar.gz)

Notes on Reproducing:
- set `rust-analyzer.checkOnSave.command` to `clippy`
- run `cargo clean` in the workspace
- restart VS Code (for some reason not restarting VS Code after changing from `check` to `clippy` makes it remember that the error exists)

**Comment by lnicola** [comment]
I seem to get the errors:

![image](https://user-images.githubusercontent.com/308347/103369822-cb0b4380-4ad3-11eb-904c-bf7a6baf7f6c.png)

However, `cargo clippy` will only show the errors the first time (unlike `cargo check`), are you sure that's not why it's working after `cargo clean`?

**Comment by lnicola** [comment]
Also, your archive isn't a `cargo` workspace because it's missing the outer `Cargo.toml`.

**Comment by lnicola** [comment]
See also https://github.com/rust-analyzer/rust-analyzer/issues/6850 for a case of disappearing errors.

**Comment by Ferdi265** [comment]
> Also, your archive isn't a `cargo` workspace because it's missing the outer `Cargo.toml`.

Oh sorry, I seem to have forgotten to add the outer `Cargo.toml`. (it is there in the tree listing I added above, but not in the archive)

Here's the fixed version: [workspace.tar.gz](https://github.com/rust-analyzer/rust-analyzer/files/5755119/workspace.tar.gz)

**Comment by lnicola** [comment]
Yeah, I made one while testing. I don't know what's happening: I expected that the error would always be shown, and that the warnings would go away quickly. Instead I get both the error and warnings in Code, but only the errors when running `cargo clippy` from the command line.

**Comment by Ferdi265** [comment]
![VS Code Problems](https://user-images.githubusercontent.com/4077106/103371017-b1b3c880-4ace-11eb-9c33-3d99523b5e6b.png)

This is what I see after running `cargo clean` and opening VS Code afterwards.

**Comment by lnicola** [comment]
What happens if you close Code and run `cargo clippy`?

**Comment by Ferdi265** [comment]
![clippy](https://user-images.githubusercontent.com/4077106/103371167-153df600-4acf-11eb-996b-ca3e996b7c15.png)

`cargo clippy` finds the error, but only that error. opening VS Code afterwards gives me the same view as before.

I think the fact that `cargo clippy` doesn't recurse into the `#[cfg(test)]` modules might be part of the reason for the different behaviour. I can try what happens when I remove the `#[cfg(test)]`.

**Comment by lnicola** [comment]
> I think the fact that cargo clippy doesn't recurse into the #[cfg(test)] modules might be part of the reason for the different behaviour. 

Right, `cargo clippy --all-targets` shows them.

**Comment by Ferdi265** [comment]
Oh, and `cargo clippy --all-targets` also has the same behaviour of not catching the `let x = y;` error.

So this might be an issue with Clippy then?

**Comment by lnicola** [comment]
> Oh, and cargo clippy --all-targets also has the same behaviour of not catching the let x = y; error.

That doesn't happen to me.

> So this might be an issue with Clippy then?

We try show the Clippy output, if an error is missing from there, we can't show it. It's an upstream bug if you can reproduce it without `rust-analyzer`.

**Comment by Ferdi265** [comment]
> We try show the Clippy output, if an error is missing from there, we can't show it. It's an upstream bug if you can reproduce it without `rust-analyzer`.

right! thanks, I'll look into it!

**Comment by lnicola** [comment]
Let's close this, but feel free to link to the `clippy` issue and/or reopen if necessary.

**Comment by Ferdi265** [comment]
Oh wow, I found it, and I now also know why it probably doesn't happen for you:

`cargo clippy` aborts on the first error, and cancels analysis of the other crates in the workspace on an error. If a job for one of the crates is already being checked in a different thread. If the system has enough threads to have started analysing `bug` in parallel, the `let x = y;` error will be found.

You can reproduce my behaviour with `cargo clippy --all-targets -j4`, and I can reproduce the other behaviour with `cargo clippy --all-targets -j8`.

This changes things, because now my main gripe is that `deny` lints are treated as a hard error, including the fact that it cancels analysis, even though that could perfectly well continue.

**Comment by lnicola** [comment]
Yeah, I was thinking it might be something like that when you mentioned the long dependency chain. It's even `-j32` in my case :-).

> This changes things, because now my main gripe is that deny lints are treated as a hard error, including the fact that it cancels analysis, even though that could perfectly well continue.

With `cargo check`, not even errors stop the compilation, so that sounds like a reasonable gripe.
